src/main/java/ui/components/pickers/LabelPickerUILogic.java

```diff
@@ -59,58 +71,146 @@ private void populatePanes() {
     }
 
-    public void toggleSelectedLabel(String text) {
-        if (!bottomLabels.isEmpty() && !text.isEmpty() && hasHighlightedLabel()) {
-            toggleLabel(
-                    bottomLabels.stream().filter(PickerLabel::isHighlighted).findFirst().get().getActualName());
-        }
+    public ArrayList<String> toggleHighlightedLabel() {
+        return toggleLabel(getHighlightedLabelName().get().getActualName());
     }
 
-    @SuppressWarnings("PMD")
+    @SuppressWarnings("unused")
+    private void ______TEXT_FIELD______() {}
+
     public void processTextFieldChange(String text) {
-        String[] textArray = text.split(" ");
-        if (textArray.length > 0) {
-            String query = textArray[textArray.length - 1];
-            if (previousNumberOfActions != textArray.length || !query.equals(lastAction)) {
-                previousNumberOfActions = textArray.length;
-                lastAction = query;
-                boolean isBottomLabelsUpdated = false;
-
-                // group check
-                // TODO rewrite this to remove some nesting, and the PMD warning
-                if (TurboLabel.getDelimiter(query).isPresent()) {
-                    String delimiter = TurboLabel.getDelimiter(query).get();
-                    String[] queryArray = query.split(Pattern.quote(delimiter));
-                    if (queryArray.length == 1) {
-                        isBottomLabelsUpdated = true;
-                        updateBottomLabels(queryArray[0], "");
-                    } else if (queryArray.length == 2) {
-                        isBottomLabelsUpdated = true;
-                        updateBottomLabels(queryArray[0], queryArray[1]);
-                    }
-                }
+        if (isEmpty(text)) {
+            removePossibleLabel();
+            restoreBottomLabels();
+        } else {
+            processChangeAndUpdateVariables(text);
+        }
+        populatePanes();
+    }
 
-                if (!isBottomLabelsUpdated) {
-                    updateBottomLabels(query);
-                }
+    private void processChangeAndUpdateVariables(String text) {
+        if (isLongerTextFieldThanPrevious(text)) {
+            addCharHandler(text);
+        } else {
+            removeCharHandler(text);
+        }
 
-                if (hasHighlightedLabel()) {
-                    addRemovePossibleLabel(getHighlightedLabelName().get().getActualName());
-                } else {
-                    addRemovePossibleLabel("");
-                }
-                populatePanes();
-            }
+        updateVariables(text);
+    }
+
+    /**
+     * This method will either toggle the highlighted label, or update both the bottom labels and the possible label
+     * in the top pane, depending on the last character of text (added character)
+     */
+    private void addCharHandler(String text) {
+        if (isLastCharSpace(text)) {
+            // space-toggle
+            spaceToggle();
```

Comment:
It toggles the highlighted label if it the last character is a space. Handled differently from a normal toggle (clicking) as label history has to be stored to support backspace.